### PR TITLE
GH-43967: [C++] Enhance error message for URI parsing

### PR DESCRIPTION
**Pull request by crystalxyz**
### Rationale for this change

We want to enhance error message for URI parsing error to provide more information for the syntax error scenario.

When error message is generated from `uriParseSingleUriExA`, the return value might indicate a `URI_ERROR_SYNTAX` error, and `error_pos` would be set to the position causing syntax error. ([uriparser/Uri.h](https://github.com/apache/arrow/blob/c455d6b8c4ae2cb22baceb4c27e1325b973d39e1/cpp/src/arrow/vendored/uriparser/Uri.h#L288))

In the new error message, it includes the character causing syntax error and its position, so users can have a better idea why the error happens.

### What changes are included in this PR?

- Error message change in URI parsing function.

### Are these changes tested?

PR includes unit tests.

### Are there any user-facing changes?

Yes, but only for error message.

* GitHub Issue: apache/arrow#41365
* GitHub Issue: apache/arrow#43967
* GitHub Issue: #43967

### CODE REVIEW

File: cpp/src/arrow/filesystem/filesystem_test.cc [hunk 1]
```diff
@@ -632,6 +633,13 @@ TEST_F(TestMockFS, FileSystemFromUri) {
   ASSERT_OK_AND_ASSIGN(fs_, FileSystemFromUri("mock:///foo/bar?q=zzz", &path));
   ASSERT_EQ(path, "foo/bar");
   CheckDirs({});
+  ASSERT_OK_AND_ASSIGN(fs_, FileSystemFromUri("mock:/folder+name/bar?q=zzz", &path));
```
Comment (AlenkaF) [inline cpp/src/arrow/filesystem/filesystem_test.cc:636]: Should `+` in the path raise or is that expected for white space? In any case we should add URIs with white space to the test as this was reported in the issue.

**Review by AlenkaF — COMMENTED** [review]
Thank you for working on this fix @CrystalZhou0529 !

From a quick look the change in cpp/src/arrow/util/uri.cc LGTM. I do have some questions/suggestions for the tests.

First: could we also add a test on the Python side to check it works correctly there also?

File: cpp/src/arrow/filesystem/filesystem_test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/filesystem_test.cc:636]: Well, there _is_ a test with whitespace below :-)

**Review by pitrou — APPROVED** [review]
This is a neat improvement, thanks @CrystalZhou0529 .

**Comment by pitrou** [comment]
@AlenkaF 

> First: could we also add a test on the Python side to check it works correctly there also?

I think the problem is that PyArrow swallows the URI error and falls back to the local filesystem (which is often legitimate, but probably not if the user really meant to pass a S3 URI :-)).

File: cpp/src/arrow/filesystem/filesystem_test.cc (hunk 1 above)
Comment (AlenkaF) [reply cpp/src/arrow/filesystem/filesystem_test.cc:636]: Oh 🤦‍♀️  sorry about that :D

**Comment by AlenkaF** [comment]
> I think the problem is that PyArrow swallows the URI error and falls back to the local filesystem (which is often legitimate, but probably not if the user really meant to pass a S3 URI :-)).

Yeah, the content in the issue should have given me this explanation already. Thanks again ;)

The Java failure is something I noticed on other PRs also if I am not mistaken?

**Review by AlenkaF — APPROVED** [review]

**Comment by pitrou** [comment]
> The Java failure is something I noticed on other PRs also if I am not mistaken?

Yes, it's quite recurring. I'm not sure there's an issue open for it?

**Comment by AlenkaF** [comment]
I think this might be the open issue connected to the Java failures: https://github.com/apache/arrow/issues/43576

Will merge this PR as the CI is green.

**Comment by vibhatha** [comment]
@AlenkaF yes this is already tracked and I checked the CIs too. And not relevant to this PR.

**Comment by jorisvandenbossche** [comment]
Thanks @CrystalZhou0529!

**Pull request merged**